lib/core/src/text/buffer.rs

```diff
@@ -0,0 +1,153 @@
+use crate::text::Line;
+use crate::text::font::FontRenderInfo;
+
+use basegl_backend_webgl::Context;
+use js_sys::Float32Array;
+use nalgebra::Transform2;
+use web_sys::WebGlBuffer;
+
+
+// =============================
+// === TextComponentsBuffers ===
+// =============================
+
+/// A structure managing all WebGl buffers used by TextComponent
+///
+/// Each attribute buffer is split to equal-sized fragments, and each fragment may is assigned to
+/// displayed line The fragment keeps the data for this line.
+#[derive(Debug)]
+pub struct TextComponentBuffers {
+    pub vertex_position     : WebGlBuffer,
+    pub texture_coordinates : WebGlBuffer,
+    fragments               : Vec<BufferFragment>,
+    pub displayed_lines     : usize,
+    pub displayed_columns   : usize,
+}
+
+/// A buffer fragment which may be assigned to some line.
+#[derive(Debug)]
+struct BufferFragment {
+    assigned_line              : Option<usize>
+    // TODO [AO] maybe some new fields appear in future. if not, remove this and use Option<usize>
+    // directly
+}
+
+impl TextComponentBuffers {
+    /// Create and initialize buffers.
+    pub fn new(gl_context:&Context, displayed_lines:usize, displayed_columns:usize)
+    -> TextComponentBuffers {
+        TextComponentBuffers {
+            vertex_position     : gl_context.create_buffer().unwrap(),
+            texture_coordinates : gl_context.create_buffer().unwrap(),
+            fragments           : Self::build_fragments(displayed_lines),
+            displayed_lines,
+            displayed_columns,
+        }
+    }
+
+    fn build_fragments(displayed_lines:usize) -> Vec<BufferFragment> {
+        let indexes   = 0..displayed_lines;
+        let fragments = indexes.map(|_| BufferFragment { assigned_line : None });
+        fragments.collect()
+    }
+
+    /// Refresh the whole buffers making them display the given lines.
+    pub fn refresh_all<'a>(&mut self, gl_context:&Context, lines:&[Line], font:&'a mut FontRenderInfo, to_window:&Transform2<f64>) {
+        self.assign_fragments(lines.len());
+
+        let content_from_index = |index : Option<usize>| index.map_or("", |i| lines[i].content.as_str());
+
+        let assigned_lines     = self.fragments.iter().map(|fragment| fragment.assigned_line);
+        let content            = assigned_lines.clone().map(content_from_index);
+        let line_indexes       = assigned_lines.map(|index| index.unwrap_or(0));
+        let content_with_index = content.clone().zip(line_indexes);
+
+        self.refresh_vertex_position_buffer(gl_context, content_with_index,font,to_window);
+        self.refresh_texture_coordinates_buffer(gl_context, content,font);
+    }
+
+    fn refresh_vertex_position_buffer<'a, Iter>
+    ( &self
+    , gl_context:&Context
+    , content_with_index:Iter
+    , font:&mut FontRenderInfo
+    , to_window:&Transform2<f64>
+    ) where Iter : ExactSizeIterator<Item=(&'a str,usize)>
+    {
```

Comment:
Just styling note but imo important - lets always use open curly bracese in previous line please!